lib/Conversion/TosaToXTenNN.cpp

```diff
@@ -0,0 +1,225 @@
+        {}, IntegerType::get(rewriter.getContext(), 8,
+                             IntegerType::SignednessSemantics::Signed));
+
+    auto newQuantizeOp = rewriter.create<amd::xten_nn::QuantizeOp>(
+        quantizeOp->getLoc(), newOutputType, quantizeOp->getOperand(0),
+        /*shift*/ (int32_t)0);
+    rewriter.replaceOpWithNewOp<amd::xten_nn::DequantizeOp>(
+        castOp, outputTensorType, newQuantizeOp->getResult(0),
+        /*shift*/ (int32_t)0);
+    rewriter.eraseOp(quantizeOp);
+
+    return success();
+  }
+};
+
+/// Folds multiplications surrounding a QDQ pair into the operations if the
+/// constants on the multiplications are power-of-two values and are equal.
+/// We assume the first multiplication represents the quantize meaning 1/scale
+/// and the second multiplication the dequantize or simply scale.
+class FoldMulsToQDQOps : public OpRewritePattern<tosa::MulOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::MulOp mulOp,
+                                PatternRewriter &rewriter) const override {
+
+    // The multiplication should have only a single constant value
+    APFloat dequantizeScaleFactor(0.0);
+    if (!m_ConstantFloat(&dequantizeScaleFactor)
+             .match(mulOp.getOperand(1).getDefiningOp())) {
+      return rewriter.notifyMatchFailure(mulOp.getOperand(1).getLoc(),
+                                         "expected to be a constant.");
+    }
+
+    // We want to make sure we have QDQ operations surrounded by MULs.
+    Operation *mulOperandOp = mulOp->getOperand(0).getDefiningOp();
+    APFloat quantizeScaleFactor(0.0);
+    auto isQDQPattern = m_Op<amd::xten_nn::DequantizeOp>(
+        m_Op<amd::xten_nn::QuantizeOp>(m_Op<tosa::MulOp>(
+            matchers::m_Any(), m_ConstantFloat(&quantizeScaleFactor))));
+    if (mulOperandOp == nullptr || !isQDQPattern.match(mulOperandOp)) {
+      return rewriter.notifyMatchFailure(mulOp->getLoc(),
+                                         "expected mul->q->dq->mul pattern.");
+    }
+    auto dequantizeOp = cast<amd::xten_nn::DequantizeOp>(mulOperandOp);
+
+    // The quantize operation will have it's reciprocal value constant folded.
+    // So we need to calculate one over to get the scale factor back.
+    APFloat recipValue((float)1.0);
+    recipValue.divide(quantizeScaleFactor, APFloat::rmNearestTiesToEven);
+    if (recipValue != dequantizeScaleFactor) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "expected constants of both multiplications around "
+                          "the QDQ to be equal.");
+    }
+
+    // Only power of two values are supported by the QDQ operations
+    std::optional<int32_t> scaleFactor =
+        getLog2Value(dequantizeScaleFactor.convertToFloat());
+    if (!scaleFactor.has_value()) {
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "constant is not a integer log2 value.");
+    }
+
```

Comment:
You will also need to check whether the muls first operand has the same shape as the muls result. Otherwise your transformation is not legal. The shapes could be different when broadcasting happens.